src/collect/mod.rs

```diff
@@ -25,6 +25,10 @@ const PERF_DFLT_INITIAL_TRACE_BUFSIZE: size_t = 1024 * 1024; // 1MiB
 /// The interface offered by all trace collectors.
 pub trait TraceCollector: Send + Sync {
     /// Return a `ThreadTraceCollector` for tracing the current thread.
+    ///
+    /// Note that an instance of `ThreadTraceCollector` is not thread local, and an attempt to
```

Comment:
I find this really difficult to parse. Can we simplify it to "You must not start two `TraceCollector`s in the same thread. Doing so is is undefined behaviour". Also I think this function (or the trait?) should be `unsafe`.

Reply:
I was thinking yesterday: why does a user facing `ThreadTraceCollector` even exist? If we had `TraceCollector::start_thread_collector()`, then this issue (I think) disappears?  There can we a hidden thread-local `ThreadTraceCollector` and the problem described in #101 I think disappear too.

Reply:
I don't know the internal API, so I'm happy with whatever change you suggest (but I do want the comments to be understandable by those of us whose English is not as good as yours :)).

Reply:
I'm actually OK with making a slightly stronger restriction than what the (current) API allows, so how about something really simple, like:

> You may only call `thread_tracer()` once per thread. Calling it more than once per-thread will result in undefined behavior.

This will force the user to only ever have one hanging around.

Reply:
> You may only call thread_tracer() once per thread.

Does that mean that you can only trace one *ever* in the lifetime of a thread?

Reply:
Not sure what you mean. One what?

Reply:
Oops, "once ever".

In other words, can I start/stop a trace (thread_tracer?) and then, once it's stopped, start it up again? Or must I create a new thread_tracer?

Reply:
The comment above is forcing the user to re-use the same thread tracer and never create a second for subsequent tracing sessions (on that thread).

Reply:
That's not how I read it. To me it says "you can only use this thread_tracer once, ever. If you want to trace again, create another thread_tracer."

Reply:
The method `thread_collector()` (sorry, I used its old name `thread_tracer()` above) only obtains a thread collector. It does not tracing until asked to with `start_collector()`.

The comment is saying -- you may only obtain a thread collector once (but the collector is reusable).

Reply:
> you may only obtain a thread collector once (but the collector is reusable).

Let's say that :)

Reply:
I've just realised that the constraint we are documented is already violated by the test suite --  `cargo test` will use threads, and individual tests will be spinning up new instances of `ThreadTraceCollector` willy-nilly.

I think we have to roll with something like:

> A thread may obtain multiple `ThreadTraceCollector`s but must only collect a trace with one at a time.

Is that any good?

[this API is a mess -- I wish i had designed it better when we started]

Reply:
Hang on, are we changing the documentation to reflect what we *want* to happen vs. changing the documentation to reflect what *will* happen?

Reply:
> you may only obtain a thread collector once (but the collector is reusable).

overly restricts the user at the cost of being very easy to explain. In other words, there are cases where it is safe to obtain > 1 thread collector, but if we tell the user they must obtain only one, it rules all unsafe cases out.

The test suite cannot abide by these rules though.

> A thread may obtain multiple ThreadTraceCollectors but must only collect a trace with one at a time.

precisely describes when it is safe to obtain >1 thread collector.

I hope that makes sense?

Reply:
[The more I see the confusion, the more I want to fix the API]

Reply:
If we can fix the API without difficulty, that's probably more productive than trying to work out how to encode a complex condition in prose!

Reply:
I don't think it will take that long to fix, but I would like to do it as a fresh PR if that's OK.

Are you OK with putting a placeholder comment here for now if I promise that my next PR fixes the API?

We could use:

> A thread may obtain multiple ThreadTraceCollectors but must only collect a trace with one at a time.

Even if it's a little weak.

Reply:
(Actually I already have a boring docstring branch to go in after this one -- if that one could go in next, I could save time resolving conflicts later)